academy/xl-cosmjs/first-steps.md

```diff
@@ -0,0 +1,637 @@
+
+const decodedTx: Tx = Tx.decode(faucetTx.tx)
+```
+
+Next you need to deserialize the only message in it:
+
+```typescript
+// import { MsgSend } from "cosmjs-types/cosmos/bank/v1beta1/tx"
+
+const sendMessage: MsgSend = MsgSend.decode(decodedTx.body!.messages[0].value)
+```
+
+In this message, the `fromAddress` is the faucet:
+
+```typescript
+const faucet = sendMessage.fromAddress
+```
+
+You can also add a line to view the faucet's balances. When running, you should get:
+
+```
+Faucet: cosmos15aptdqmm7ddgtcrjvc5hs988rlrkze40l4q0he
+Faucet balances: [ { denom: 'uatom', amount: '867777337235' } ]
+```
+
+<ExpansionPanel title="Get the faucet address another way">
+
+If you want to experiment more, you can parse the `rawLog` (an _output_ of the computation) as opposed to deserializing the transaction proper (the _input_ of the computation).
+
+From the `IndexedTx` you see that there is a [`rawLog`](https://github.com/cosmos/cosmjs/blob/13ce43c/packages/stargate/src/stargateclient.ts#L64), which happens to be a stringified JSON.
+
+```typescript
+const rawLog = JSON.parse(faucetTx.rawLog)
+console.log(JSON.stringify(rawLog, null, 4))
+```
+
+The structure of the raw log is not always obvious, but in this example it contains:
+
+```json
+[
+    {
+        "events": [
+            ...
+            {
+                "type": "coin_spent",
+                "attributes": [
+                    {
+                        "key": "spender",
+                        "value": "cosmos15aptdqmm7ddgtcrjvc5hs988rlrkze40l4q0he"
+                    },
+                    ...
+                ]
+                ...
+            }
+            ...
+        ]
+    }
+]
+```
+
+Therefore your faucet address is:
+
+```typescript
+const faucet = rawLog[0].events[1].attributes[0].value
+```
+
+</ExpansionPanel>
+
+These actions are example uses of the read-only `StargateClient`.
+
+Now it is time for Alice to send some tokens back to the faucet.
+
+## Prepare a signing client
+
+Alice cannot send tokens with the read-only `StargateClient`. She needs a _signing client_ that lets her sign transactions. Update your import line:
+
+```typescript
+import { IndexedTx, SigningStargateClient, StargateClient } from "@cosmjs/stargate"
+```
+
+VSCode's auto-complete can assist you again. When you instantiate this new client, you need to pass it a **signer**, implementing the `OfflineDirectSigner` interface. The signer needs access to Alice's **private key**, and there are several ways to accomplish this. In this example, you will use Alice's saved mnemonic.
+
+Load the mnemonic as text in your code with this import:
+
+```typescript
+import { readFile } from "fs/promises"
+```
+
+There are several implementation of `OfflineDirectSigner` available. Right-click on `OfflineDirectSigner` in VSCode and select <kbd>Find All Implementations</kbd>: `DirectSecp256k1HdWallet` is the most appropriate for this situation. Add the import:
```

Comment:
Would be great to explain why `DirectSecp256k1HdWallet` is the most appropriate one for us.